src/features/atomics.md

```diff
@@ -0,0 +1,215 @@
+
+The `unlock` is method sets the lock to the `UNLOCKED` state and calls `Atomics.notify` to wake up one waiter which was waiting for the lock. The state change is always expected to succeed, since this thread is holding the lock, and nobody else should call `unlock()` meanwhile.
+
+```js
+unlock() {
+  const oldValue = Atomics.compareExchange(this.i32a, AsyncLock.INDEX,
+                      /* old value >>> */  AsyncLock.LOCKED,
+                      /* new value >>> */  AsyncLock.UNLOCKED);
+  if (oldValue != AsyncLock.LOCKED) {
+    throw new Error("Tried to unlock while not holding the mutex");
+  }
+  Atomics.notify(this.i32a, AsyncLock.INDEX, 1);
+}
+```
+
+The straightforward case goes as follows: the lock is free and thread T1 acquires it by changing the lock state with `Atomics.compareExchange`. Thread T2 tries to acquire the lock by calling `Atomics.compareExchange`, but it doesn’t succeed in changing the lock state. T2 then calls `Atomics.wait`, which blocks the thread. At some point T1 releases the lock and calls `Atomics.notify`. That makes the `Atomics.wait` call in T2 return `"ok"`, waking up T2. T2 then tries to acquire the lock again, and this time succeeds.
+
+There are also 2 possible corner cases - these demonstrate the reason for `Atomics.wait` and `Atomics.waitAsync` checking for a specific value at the index:
+
+- T1 is holding the lock and T2 tries to get it. First, T2 tries to change the lock state with `Atomics.compareExchange`, but doesn't succeed. But then T1 releases the lock before T2 manages to call `Atomics.wait`. When T2 calls `Atomics.wait`, it will return immediately with the return value `"not-equal"`. In that case, T2 will continue with the next loop iteration, trying to acquire the lock again.
+- T1 is holding the lock and T2 is waiting for it with `Atomics.wait`. T1 releases the lock - T2 wakes up (the `Atomics.wait` call returns) and tries to do `Atomics.compareExchange` to acquire the lock, but another thread T3 was faster and got the lock already. So the call to `Atomics.compareExchange` fails to get the lock, and T2 calls `Atomics.wait` again, blocking until T3 releases the lock.
+
+Because of the latter corner case, the mutex isn’t "fair". It’s possible that T2 has been waiting for the lock to be released, but T3 comes and gets it immediately. A more realistic lock implementation may use several states to differentiate between "locked" and "locked with contention".
+
+## Async lock
+
+The non-blocking `executeLocked` method is callable from the main thread, unlike the blocking `lock` method. It gets a callback function as its only parameter and schedules the callback to be executed once it has successfully acquired the lock.
+
+```js
+executeLocked(f) {
+  const self = this;
```

Comment:
We could probably use an arrow function for `tryGetLock` instead. Not sure if clearer though.

Reply:
Yea I think this is clearer.